applications/incompressible_navier_stokes/navier_stokes_manufactured/application.h

```diff
@@ -605,28 +614,51 @@ class Application : public ApplicationBase<dim, Number>
-            typename dealii::Triangulation<dim>::cell_iterator>> & periodic_face_pairs,
-          unsigned int const                                       global_refinements,
-          std::vector<unsigned int> const &                        vector_local_refinements) {
-        (void)periodic_face_pairs;
-
+    auto const lambda_create_triangulation = [&](dealii::Triangulation<dim, dim> & tria,
+                                                 std::vector<dealii::GridTools::PeriodicFacePair<
+                                                   typename dealii::Triangulation<
+                                                     dim>::cell_iterator>> & periodic_face_pairs,
+                                                 unsigned int const          global_refinements,
+                                                 std::vector<unsigned int> const &
+                                                   vector_local_refinements) {
+      (void)vector_local_refinements;
+
+      if(boundary_condition == BoundaryCondition::Periodic)
+      {
+        AssertThrow(
+          this->param.grid.triangulation_type != TriangulationType::FullyDistributed,
+          dealii::ExcMessage(
+            "Periodic faces might not be applied correctly for TriangulationType::FullyDistributed. "
+            "Try to use another triangulation type, or try to fix these limitations in ExaDG or deal.II."));
+
+        bool const is_symmetric_around_origin = std::abs(interval_start + interval_end) < 1e-18;
+        bool const half_length_is_pi = std::abs(interval_end - dealii::numbers::PI) < 1e-18;
+
+        AssertThrow(is_symmetric_around_origin && half_length_is_pi,
+                    dealii::ExcMessage(
+                      "Solution is only periodic on [-pi, pi]^2 domain (ignoring multiples)."));
+
+        create_periodic_box(tria,
+                            0 /* n_refine_space */,
+                            periodic_face_pairs,
+                            1 /* repetitions */,
+                            interval_start,
+                            interval_end);
+      }
+      else
+      {
         dealii::GridGenerator::hyper_cube(tria, interval_start, interval_end, true);
+      }
 
-        // Save the *coarse* triangulation for later deserialization.
-        if(write_restart and this->param.grid.triangulation_type == TriangulationType::Serial)
-        {
-          save_coarse_triangulation<dim>(this->param.restart_data, tria);
-        }
-
-        if(vector_local_refinements.size() > 0)
-          refine_local(tria, vector_local_refinements);
+      // Save the *coarse* triangulation for later deserialization.
+      if(write_restart and this->param.grid.triangulation_type == TriangulationType::Serial)
+      {
+        save_coarse_triangulation<dim>(this->param.restart_data, tria);
+      }
 
-        if(global_refinements > 0)
-          tria.refine_global(global_refinements);
-      };
+      if(global_refinements > 0)
```

Comment:
we refine afterwards, since we have to store the coarse tria in case of restart. I want to write some tests for restart next.